Modules/_testcapi/file.c

```diff
@@ -57,9 +59,61 @@ _testcapi_py_fopen_impl(PyObject *module, PyObject *path, const char *mode,
 }
 
 
+/*[clinic input]
+_testcapi.py_universalnewlinefgets
+
+    file: object
+    size: int
+    /
+
+Read a line from a file using Py_UniversalNewlineFgets.
+[clinic start generated code]*/
+
+static PyObject *
+_testcapi_py_universalnewlinefgets_impl(PyObject *module, PyObject *file,
+                                        int size)
+/*[clinic end generated code: output=2ce1bc76c9dc871c input=02c236049d18569a]*/
+{
+    int fd = PyObject_AsFileDescriptor(file);
+    if (fd == -1) {
+        return NULL;
+    }
+
+    FILE *fp;
+#ifdef MS_WINDOWS
+    fp = _fdopen(fd, "rb");
+#else
+    fp = fdopen(fd, "rb");
+#endif
```

Comment:
You may pass a filename and call `Py_fopen(file)` instead to avoid having to care about *fd*. So calling `Py_fclose(fp)` would be safe in all cases since we own the internal file descriptor.